File: 313.Super.Ugly.Number.py
class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        m = len(primes)
        f = [0] * m
        ret = [1]
        while len(ret) < n:
            mi = float('inf')
            for i in range(m):
                num = ret[f[i]]*primes[i]
                while num <= ret[-1]:
                    f[i] += 1
                    num = ret[f[i]]*primes[i]
                mi = min(mi, num)
            ret.append(mi)
        return ret[n-1]
    

File: 313. Super Ugly Number.go
func nthSuperUglyNumber(n int, primes []int) int {
	ugly := make([]int, n)
	ugly[0] = 1
	index := make([]int, len(primes))
	nums := make([]int, len(primes))
	min := 0
	for i := 1; i < n; i++ {
		min = math.MaxInt32
		for j := 0; j < len(primes); j++ {
			nums[j] = ugly[index[j]] * primes[j]
			if nums[j] < min {
				min = nums[j]
			}
		}
		ugly[i] = min
		for j := 0; j < len(primes); j++ {
			if nums[j] == min {
				index[j]++
			}
		}
	}
	return ugly[n-1]
}